API::Object allocations leak in ObjC ARC due to internal strong wrapper reference.

xeenon opened the pull request:
#### bbb9a28e95477f7f83bb0b2b822d5aebfecf6b44
<pre>
API::Object allocations leak in ObjC ARC due to internal strong wrapper reference.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248205">https://bugs.webkit.org/show_bug.cgi?id=248205</a>

Reviewed by Darin Adler and David Kilzer.

The m_wrapper in `API::Object` was counting as a strong reference in objects using ObjC ARC.
This was affecting all the new `_WKWebExtension*` API classes.

Changed `m_wrapper` in `API::Object` to be a `CFTypeRef` instead. This hides the references from
ARC and continues to let `API::Object` manage the ref counting with `CFRetain`/`CFRelease`.

Tested with automated WKWebExtension API tests and manual testing with Xcode&apos;s memory graph.

* Source/WebKit/Shared/API/APIObject.h:
(API::Object::constructInWrapper): Bridge cast to `CFTypeRef` when setting m_wrapper. Use `id &lt;WKObject&gt;`.
(API::Object::wrapper const): Changed to return `id &lt;WKObject&gt;` instead of NSObject *. This
better balances with `constructInWrapper()` taking `id &lt;WKObject&gt;`. wrapper() is now also limited
to ObjC contexts, like `constructInWrapper()`, which ObjC/ObjC++ files is the only place it is used.
The `WKObject` protocol conforms to the `NSObject` protocol, so `NSObject &lt;WKObject&gt; *` is redundant.
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::ref const): Use `m_wrapper` instead of `wrapper()` since the type is now `CFTypeRef`.
(API::Object::deref const): Ditto.
(API::Object::newObject): Bridge cast to `CFTypeRef` when setting m_wrapper.
* Source/WebKit/Shared/Cocoa/WKNSArray.mm:
(-[WKNSArray objectAtIndex:]): Cast `wrapper()` to id to avoid warning about NSNull not conforming to `WKObject`.
* Source/WebKit/Shared/Cocoa/WKNSDictionary.mm:
(-[WKNSDictionary objectForKey:]): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtension.h: Wrap `wrapper()` in `#if __OBJC__` to match `API::Object`.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h: Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h: Ditto.
* Source/WebKit/UIProcess/PageClient.h: Added `OBJC_CLASS NSObject;` to fix build.

Canonical link: <a href="https://commits.webkit.org/256954@main">https://commits.webkit.org/256954@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/69c25a7baa36e9f95aa7c897556aa0d55d2adaa3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97334 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6598 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30493 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106853 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6901 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35336 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89744 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103532 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102998 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5184 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83964 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32184 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87036 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88858 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75109 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/614 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20324 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/597 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5398 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2355 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1841 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41123 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler reviewed (approved):
Does this work when some files are compiled with ARC and other files compiled without it, and they are both manipulating the same objects? I would have expected we’d need to use `WeakObjCPtr` for non-ARC too.

ddkilzer commented on Source/WebKit/Shared/API/APIObject.h:
```diff
@@ -231,14 +235,18 @@ class Object
 #ifdef __OBJC__
     template<typename T, typename... Args>
-    static void constructInWrapper(NSObject <WKObject> *wrapper, Args&&... args)
+    static void constructInWrapper(id<WKObject> wrapper, Args&&... args)
```
> To appease the compiler, also switched the types to "`id`" to avoid a "type not inherited" warning in `WeakObjCPtr.h`.

What (or how many) objects are not inheriting from `NSObject`?  That seems like a bug.  If it's only a handful, we should fix that (perhaps in a follow-up).

xeenon replied:
All of our objects inherit from `NSObject`. However, I think it was the `OBJC_CLASS NSObject` causing the confusion in the compiler.

ddkilzer commented on Source/WebKit/Shared/API/APIObject.h:
```diff
@@ -261,7 +269,11 @@ class Object
+    WeakObjCPtr<id> m_wrapper;
+#else
     NSObject *m_wrapper;
+#endif
```
> Does this work when some files are compiled with ARC and other files compiled without it, and they are both manipulating the same objects? I would have expected we’d need to use `WeakObjCPtr` for non-ARC too.

In non-ARC mode (e.g., C++ source), we are basically storing raw pointers.  We should switch both to use `WeakObjCPtr`.  It will simplify the code change as well since you can get rid of the `!__has_feature(objc_arc)` code.

xeenon replied:
I initially tried that and the `WeakObjCPtr` was zeroing out during construction in the non-ARC case. I'll dive into that some more and see if I can find a solution.

ddkilzer reviewed (approved):
r=me, but I'd really like to see us use `WeakObjCPtr` in all cases.

darinadler commented on Source/WebKit/Shared/Cocoa/WKNSArray.mm:
```diff
@@ -52,7 +52,7 @@ - (NSUInteger)count
 {
     API::Object* object = _array->at(i);
-    return object ? object->wrapper() : [NSNull null];
+    return object ? (id)object->wrapper() : [NSNull null];
```
I am surprised that the cast to `id` is needed. What kind of error do you get without it?

xeenon replied:
```
Source/WebKit/Shared/Cocoa/WKNSArray.mm:55:19:
Incompatible operand types ('id<WKObject>' and 'NSNull * _Nonnull')
```

xeenon replied:
Doing `return object ? object->wrapper() : (id)[NSNull null]` works too.

ddkilzer replied:
This needs to use a proper `(__bridge id)` cast to work with ARC later.  Please use `bridge_id_cast()` from `<wtf/cocoa/TypeCastsCocoa.h>`:

```
    return object ? bridge_id_cast(object->wrapper()) : [NSNull null];
```

I think casting `object->wrapper()` is preferred here since it's required once this file is built with ARC.

xeenon replied:
`wrapper()` is already doing a `(__bridge id <WKObject>` cast in it. This is just a secondary cast. Does that matter?

ddkilzer replied:
Oops, `wrapper()` already returns `id <WKObject>`.  Got confused scrolling up and down patch.

So this doesn't need to be changed to enable ARC later.  Using `(id)` cast is fine.

darinadler commented on Source/WebKit/Shared/Cocoa/WKNSDictionary.mm:
```diff
@@ -68,7 +68,7 @@ - (id)objectForKey:(id)key
         return nil;
 
-    return value ? value->wrapper() : [NSNull null];
+    return value ? (id)value->wrapper() : [NSNull null];
```
I am surprised that the cast to `id` is needed. What kind of error do you get without it?

xeenon replied:
Same as above with `WKNSArray.mm`.

ddkilzer commented on Source/WebKit/Shared/Cocoa/WKNSDictionary.mm:
```diff
@@ -68,7 +68,7 @@ - (id)objectForKey:(id)key
         return nil;
 
-    return value ? value->wrapper() : [NSNull null];
+    return value ? (id)value->wrapper() : [NSNull null];
```
Please use `bridge_id_cast(value->wrapper())` here.

ddkilzer replied:
Okay, `(id)` cast here is fine.

ddkilzer commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -214,8 +214,10 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
     NSArray *errors();
 
+#ifdef __OBJC__
     _WKWebExtension *wrapper() const { return (_WKWebExtension *)API::ObjectImpl<API::Object::Type::WebExtension>::wrapper(); }
```
Would be nice to use `dynamic_objc_cast<_WKWebExtension>(...)` here from `<wtf/cocoa/TypeCastsCocoa.h>`.

(Technically it's defined in `<wtf/RetainPtr.h>`, but it may move in the future.)

xeenon replied:
That caused an error since the ObjC class is a forward declaration, and the `isKindOfClass:` check needs the ObjC class header included to do the `[T class]` call.

xeenon replied:
I'll keep these as simple casts.

ddkilzer replied:
Okay.

ddkilzer commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -193,8 +193,10 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
     void webViewWebContentProcessDidTerminate(WKWebView *);
 
+#ifdef __OBJC__
     _WKWebExtensionContext *wrapper() const { return (_WKWebExtensionContext *)API::ObjectImpl<API::Object::Type::WebExtensionContext>::wrapper(); }
```
Would be nice to use `dynamic_objc_cast<_WKWebExtensionContext>(...)` here from `<wtf/cocoa/TypeCastsCocoa.h>`.

ddkilzer commented on Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
```diff
@@ -87,8 +87,10 @@ class WebExtensionController : public API::ObjectImpl<API::Object::Type::WebExte
     void sendToAllProcesses(const T& message, ObjectIdentifier<U> destinationID);
 
+#ifdef __OBJC__
     _WKWebExtensionController *wrapper() const { return (_WKWebExtensionController *)API::ObjectImpl<API::Object::Type::WebExtensionController>::wrapper(); }
```
Would be nice to use `dynamic_objc_cast<_WKWebExtensionController>(...)` here from `<wtf/cocoa/TypeCastsCocoa.h>`.

ddkilzer commented on Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h:
```diff
@@ -101,8 +101,10 @@ class WebExtensionMatchPattern : public API::ObjectImpl<API::Object::Type::WebEx
     unsigned hash() const { return m_hash; }
 
+#ifdef __OBJC__
     _WKWebExtensionMatchPattern *wrapper() const { return (_WKWebExtensionMatchPattern *)API::ObjectImpl<API::Object::Type::WebExtensionMatchPattern>::wrapper(); }
```
Would be nice to use `dynamic_objc_cast<_WKWebExtensionMatchPattern>(...)` here from `<wtf/cocoa/TypeCastsCocoa.h>`.

ddkilzer reviewed (approved):
r=me, but at minimum you need to use `bridge_id_cast(value->wrapper())` in `WKNSArray.mm` and in `WKNSDictionary.mm` to reduce the work to enable ARC later.

xeenon committed:
API::Object allocations leak in ObjC ARC due to internal strong wrapper reference.
https://bugs.webkit.org/show_bug.cgi?id=248205

Reviewed by Darin Adler and David Kilzer.

The m_wrapper in `API::Object` was counting as a strong reference in objects using ObjC ARC.
This was affecting all the new `_WKWebExtension*` API classes.

Changed `m_wrapper` in `API::Object` to be a `CFTypeRef` instead. This hides the references from
ARC and continues to let `API::Object` manage the ref counting with `CFRetain`/`CFRelease`.

Tested with automated WKWebExtension API tests and manual testing with Xcode's memory graph.

* Source/WebKit/Shared/API/APIObject.h:
(API::Object::constructInWrapper): Bridge cast to `CFTypeRef` when setting m_wrapper. Use `id <WKObject>`.
(API::Object::wrapper const): Changed to return `id <WKObject>` instead of NSObject *. This
better balances with `constructInWrapper()` taking `id <WKObject>`. wrapper() is now also limited
to ObjC contexts, like `constructInWrapper()`, which ObjC/ObjC++ files is the only place it is used.
The `WKObject` protocol conforms to the `NSObject` protocol, so `NSObject <WKObject> *` is redundant.
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::ref const): Use `m_wrapper` instead of `wrapper()` since the type is now `CFTypeRef`.
(API::Object::deref const): Ditto.
(API::Object::newObject): Bridge cast to `CFTypeRef` when setting m_wrapper.
* Source/WebKit/Shared/Cocoa/WKNSArray.mm:
(-[WKNSArray objectAtIndex:]): Cast `wrapper()` to id to avoid warning about NSNull not conforming to `WKObject`.
* Source/WebKit/Shared/Cocoa/WKNSDictionary.mm:
(-[WKNSDictionary objectForKey:]): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtension.h: Wrap `wrapper()` in `#if __OBJC__` to match `API::Object`.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h: Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h: Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionMatchPattern.h: Ditto.
* Source/WebKit/UIProcess/PageClient.h: Added `OBJC_CLASS NSObject;` to fix build.

Canonical link: https://commits.webkit.org/256954@main
